### ISSUE TITLE
Character '-' not accepted in JSON key

### ISSUE THREAD

**Issue by git-blame** [graphile/postgraphile#1532]
### Summary

Using a mutation to add data to my database. One of the field is a JSON structure. If within that JSON structure, a key contains '-' (e.g., "user-state"), postgraphile reports `Syntax Error: Invalid number, expected digit but got: \"s\"`

If I replace with "user_state" in the example, then everything works. Unfortunately, for legacy purposes, the source of this data cannot be modified.

### Steps to reproduce

An example input:
```json
{ "query": "mutation {
  addEvent(input: {
    clientMutationId: \"AddEvent\",
    user: \"1234\",
    ip: \"127.0.1.1\",
   log: {version: 1, agent: \"monitor\", host: \"localhost\", type: \"system\", timestamp: \"1632406094\", event: {user-state: \"idle\"} }
}) {
 clientMutationId
}

}"
}
```

### Expected results
I've looked at various answers online as well as json.org. As far as I know, '-' is a valid character in a JSON key.

### Actual results

```json
{
  "errors": [
    {
      "message": "Syntax Error: Invalid number, expected digit but got: \"s\".",
      "locations": [
        {
          "line": 1,
          "column": 206
        }
      ],
      "stack": "GraphQLError: Syntax Error: Invalid number, expected digit but got: \"s\".\n    at syntaxError (/node_modules/graphql/error/syntaxError.js:15:10)\n    at readDigits (/node_modules/graphql/language/lexer.js:426:38)\n    at readNumber (/node_modules/graphql/language/lexer.js:373:16)\n    at readToken (/node_modules/graphql/language/lexer.js:259:14)\n    at Object.lookahead (/node_modules/graphql/language/lexer.js:54:43)\n    at Object.advanceLexer [as advance] (/node_modules/graphql/language/lexer.js:44:33)\n    at Parser.expectToken (/node_modules/graphql/language/parser.js:1399:19)\n    at Parser.parseName (/node_modules/graphql/language/parser.js:94:22)\n    at Parser.parseObjectField (/node_modules/graphql/language/parser.js:579:21)\n    at Parser.item (/node_modules/graphql/language/parser.js:563:21)"
    }
  ]
}
```

### Additional context

postgraphile v4.12.3

**Comment by benjie** [comment]
What you have here is not JSON - the most obvious giveaway to this is that they keys do not have quote marks around them, which is required by JSON. You are attempting to specify a JSON scalar via the GraphQL SDL, which has a very tight restriction on the keys allowed - namely they must conform to `Name`: https://spec.graphql.org/draft/#Name

There are multiple solutions you can have from here; the one I would recommend is to specify the JSON via a variable:

```json
{
  "query": "mutation AddEvent($input: AddEventInput!) { addEvent(input: $input) { clientMutationId } }",
  "variables": {
    "input": {
      "clientMutationId": "AddEvent",
      "user": "1234",
      "ip": "127.0.1.1",
      "log": {
        "version": 1,
        "agent": "monitor",
        "host": "localhost",
        "type": "system",
        "timestamp": "1632406094",
        "event": { "user-state": "idle" }
      }
    }
  }
}
```

An alternative solution would be to not use the `--dynamic-json` option and instead have your JSON scalars representable via a string. In which case you'd pass the field a GraphQL string value which contained the stringified JSON object.

**Comment by git-blame** [comment]
Thanks, this is very helpful. I was thrown off by my version of graphiql. When I was constructing my query by using it, I placed the data into the query variables. For the json field, using quotations (e.g., `"log": { "version": 1 ...`), the UI would highlight the section and complained that `Expected value of type "JSON"`.

So I moved the data directly into the query itself. But with a JSON scalar, the keys are not quoted. That seems to work except for a key such as `user-state`.

Your suggestion works great and simplifies my graphql event processing. For one thing, I don't have to remove the quotes around the keys and escape the quotes around the values. I can simply append the JSON data directly to the variables section.
